openlibrary/catalog/marc/parse.py

```diff
@@ -326,51 +321,65 @@ def read_languages(rec, lang_008: Optional[str] = None):
     return [lang_map.get(code, code) for code in found]
 
 
-def read_pub_date(rec):
+def read_pub_date(rec: MarcBase) -> str | None:
     fields = rec.get_fields('260')
-    if not fields:
-        return
     found = []
     for f in fields:
-        found += [v for v in f.get_subfield_values('c') if v]
+        found += f.get_subfield_values('c')
     return remove_trailing_number_dot(found[0].strip('[]')) if found else None
 
 
-def read_publisher(rec):
-    fields = rec.get_fields('260') or rec.get_fields('264')[:1]
+def read_publisher(rec: MarcBase) -> dict[str, Any] | None:
+    fields = (
+        rec.get_fields('260')
+        or rec.get_fields('264')[:1]
+        or [rec.get_linkage('260', '880')]
```

Comment:
Placate mypy by removing Nones...
```suggestion
        or [link for link in rec.get_linkage('260', '880') if link]
```

Reply:
Thank you @cclauss ! -- the more I looked at the error msg on this one, the less I understood the problem.

 ```or [link for link in [rec.get_linkage('260', '880')] if link]```

is what works here -- that's not something I'd normally write, but it does placate mypy. Is there a way to simplify that?

... although I think this illustrates something I was suspecting of mypy -- it doesn't handle moderately complex logical expressions well. I couldn't find a good treatment of the problem. I think it would make sense to have some correctly typed default result to convince mypy that an intermediate falsey result will never be returned, but I'm not sure the best way to do that. Activating that default would actually indicate an unexpected error had occurred.

Reply:
Woohoo, all tests passing, thank you @cclauss

Reply:
✅  🥳 Mypy is a retrofit but is better than not having type hints at all.  It does kinda force us to write code that is a bit unnatural but its intentions are good.  In general, it dislikes ([for good reasons](https://github.com/internetarchive/openlibrary/pull/6578#issuecomment-1137181218)) functions that can return multiple different data types.

It may be useful to add strict wrapper functions that return values of a single type and raise errors on other types.
```python
 def read_fields_as_marc_field_base(self, want: list[str]) -> Iterator[MarcFieldBase]]:
    for marc_field_base in self.read_fields(want):
        assert isinstance(marc_field_base, MarcFieldBase)
        yield marc_field_base
```
Happy typing...